storage/storage_image.go

```diff
@@ -0,0 +1,430 @@
+		image, err := s.store.GetImage(s.ID)
+		if image != nil && err == nil {
+			json.Unmarshal([]byte(image.Metadata), s)
+		}
+	}
+}
+
+func (s *storageImage) saveMetadata() {
+	if s.ID != "" {
+		if metadata, err := json.Marshal(s); len(metadata) != 0 && err == nil {
+			istore, err := s.store.GetImageStore()
+			if istore != nil && err == nil {
+				istore.SetMetadata(s.ID, string(metadata))
+			}
+		}
+	}
+}
+
+func (s *storageImage) Reference() types.ImageReference {
+	s.ID = s.imageRef.ID()
+	return s.imageRef
+}
+
+func (s *storageImage) Close() {
+}
+
+// SupportsSignatures returns an error if we can't expect GetSignatures() to
+// return data that was previously supplied to PutSignatures().
+func (s *storageImage) SupportsSignatures() error {
+	return nil
+}
+
+// PutBlob is used to both store filesystem layers and binary data that is part
+// of the image.
+func (s *storageImage) PutBlob(stream io.Reader, digest string, expectedSize int64) (actualDigest string, actualSize int64, err error) {
+	// Try to read an initial snippet of the blob.
+	header := make([]byte, 10240)
+	n, err := stream.Read(header)
+	if err != nil && err != io.EOF {
+		return "", -1, err
+	}
+	// Set up to read the whole blob (the initial snippet, plus the rest)
+	// while digesting it with sha256.
+	hasher := sha256.New()
+	defragmented := io.MultiReader(bytes.NewBuffer(header[:n]), stream)
+	multi := io.TeeReader(defragmented, hasher)
+	if (n > 0) && archive.IsArchive(header[:n]) {
+		// It's a filesystem layer.  If it's not the first one in the
+		// image, we assume that the most recently added layer is its
+		// parent.
+		parentLayer := ""
+		if len(s.BlobList) > 0 {
+			for _, blob := range s.BlobList {
+				if layerID, ok := s.Layers[blob]; ok {
+					parentLayer = layerID
+				}
+			}
+		}
+		// Now try to figure out if the identifier we have is an ID or
+		// something else, so that we can do collision detection
+		// correctly.
+		names := []string{digest}
+		id := digest
+		if matches := idRegexp.FindStringSubmatch(digest); len(matches) > 1 {
+			id = matches[len(matches)-1]
+			names = nil
+		}
+		layer, err := s.store.CreateLayer(id, parentLayer, names, "", true)
+		if err != nil && err != storage.ErrDuplicateID {
+			return "", -1, err
+		}
+		if err == storage.ErrDuplicateID {
+			// There's already a layer with the same ID.  If its
+			// checksum is the same as ours, then we're cool with
+			// just using the already-stored copy.
+			io.Copy(ioutil.Discard, multi)
+			hash := hasher.Sum(nil)
+			hasher = sha256.New()
+			diff, _, err := s.diffLayer(id, false)
+			if err != nil {
+				logrus.Debugf("blob %q already present but got %v while verifying it", err)
+				return "", -1, err
+			}
+			io.Copy(hasher, diff)
+			diff.Close()
+			if bytes.Compare(hash, hasher.Sum(nil)) != 0 {
+				// They are not the same.  Return an error.
+				logrus.Debugf("blob %q already present with different contents", id)
+				return "", -1, storage.ErrDuplicateID
+			}
```

Comment:
This is what https://github.com/containers/image/pull/63/files/c111f3ed169f1f5db19fc5eddd5c376387ae6491#r78212929 refers to—if we already have a copy of the layer and we _know_ that it matches the digest, we don’t really care about the input stream not matching.

OTOH I haven’t yet read all of this PR and I may be misunderstanding how the layer IDs work.